### ISSUE TITLE
New GEO Extension with commonly needed functions for GIS, geography, etc.

### ISSUE THREAD

**Issue by thadguidry** [OpenRefine/OpenRefine#6570]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%24100%20Funded-%2300A156.svg" />](https://oss.issuehunt.io/r/OpenRefine/OpenRefine/issues/6570)

[In the forum](https://forum.openrefine.org/t/how-to-use-your-own-python-scripts-as-apis-in-openrefine/1273) there seems to be some interest in OpenRefine perhaps bundling a GEO extension with various GREL functions (perhaps menu also in the future?) for performing conversions, great circle distance calculations, etc.)

### Proposed solution
Add a new GEO Extension to OpenRefine that would have a few functions that are useful.
Initially, perhaps the following kind of functions:
1. `decToGMS()` which converts decimal coordinates into GMS (degrees, minutes, seconds)
2. Great Circle Distance calculator that takes 2 coordinates (decimal, GMS, ???) and returns the great circle distance between them.

The implementations might use some existing Java libraries such as:
- [code like this](https://github.com/eugenp/tutorials/blob/master/core-java-modules/core-java-lang-math-3/src/main/java/com/baeldung/geocoordinatevalidator/GeoCoordinateValidator.java)
- or bits of [Apache Sedona libraries](https://github.com/apache/sedona/blob/master/common/src/main/java/org/apache/sedona/common/Functions.java))

it also may be worth reviewing this existing (but old) extension https://github.com/ryanfb/georefine to see if any aspects of this can be reused / updated to work with the current OpenRefine release and extension mechanisms we have currently in 3.x versions.

**Please ask questions prior to development,** so that the team agrees with the complete minimum scope and version of OpenRefine to support initially. (we have an upcoming release, but current is 3.7.9)

### Additional context

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [<img src='https://avatars.githubusercontent.com/u/986438?v=4' alt='thadguidry' width=24 height=24> thadguidry](https://oss.issuehunt.io/u/thadguidry) ($100.00)

### Submitted pull Requests
- [#7432 feat: implement GEO extension with decToGMS and geoDistance functions](https://oss.issuehunt.io/r/OpenRefine/OpenRefine/pull/7432)
---

### Tips

- Checkout the [Issuehunt explorer](https://oss.issuehunt.io/r/OpenRefine/OpenRefine/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://oss.issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by ostephens** [comment]
It may be worth reviewing this existing (but old) extension to see if any aspects of this can be reused / updated to work with the current OpenRefine release https://github.com/ryanfb/georefine

**Comment by dino2580** [comment]
I am willing to work on this issue 
According to my understanding this issue requires decToGms() and Great distance Calulator(For this i have create my own or use some already existing libraries) And what could other useful features .What problems I could due to the updated version of Openrefine

**Comment by SashankBhamidi** [comment]
Hi @thadguidry, I've implemented the GEO extension as requested in PR #7432.

The implementation provides both requested functions:
- \: Converts decimal degrees to DMS format with directional indicators  
- \: Great circle distance using Haversine formula

Implementation includes 24 comprehensive tests, follows OpenRefine standards, and uses only built-in Java libraries. Ready for review.

**Comment by SashankBhamidi** [comment]
I realize I should have asked questions before development as you mentioned. Since the requirements seemed clear, I went ahead with implementation, but want to confirm the scope:

**Questions:**
1. Does targeting OpenRefine 3.10 (current master) work for the upcoming release?
2. Is the two-function scope (decToGMS + geoDistance) sufficient for the minimum viable implementation?
3. Any concerns with using zero external dependencies vs. Apache Sedona libraries as originally suggested?

Happy to adjust the implementation if needed before review.

**Comment by thadguidry** [comment]
@tfmorris or @SoryRawyer could answer those questions for you @SashankBhamidi

**Comment by SashankBhamidi** [comment]
@SoryRawyer or @tfmorris, could one of you take a look at the PR and respond to the questions? Thanks!

**Comment by SoryRawyer** [comment]
> 1. Does targeting OpenRefine 3.10 (current master) work for the upcoming release?

Yes, I we're still wrapping things up for 3.10 so that would work.

> 2. Is the two-function scope (decToGMS + geoDistance) sufficient for the minimum viable implementation?

As long as those functions fulfill the scope in the issue, I think that's a good place to start.

> 3. Any concerns with using zero external dependencies vs. Apache Sedona libraries as originally suggested?

I'll never complain about having fewer dependencies : )

I'll post my follow up comments on the PR itself. In the meantime, thank you for contributing!

**Comment by SashankBhamidi** [comment]
Thanks @SoryRawyer!

The IssueHunt bounty is tied to this specific PR being merged in this repo (tied to the issue on issuehunt). Since you've confirmed the implementation works and issue #6570 didn't specify repository structure, I need this PR to merge for the bounty payout.

I'm happy to create a separate repo afterward and modify this to just handle integration. Let me know how you'd like to coordinate the transition to separate repo once this merges.

**Comment by tfmorris** [comment]
It looks like @ostephens [suggestion](https://github.com/OpenRefine/OpenRefine/issues/6570#issuecomment-2079048858) 

> It may be worth reviewing this existing (but old) extension to see if any aspects of this can be reused / updated to work with the current OpenRefine release https://github.com/ryanfb/georefine

got lots of thumbs ups, but did anyone look at that extension? Someone complained a few years ago that it no longer worked and Ryan fixed it, so I suspect it probably still works today and it appears to be much more comprehensive than the two functions listed above.

It includes:

> - point: creates a point from latitude and longitude in the WGS84 coordinate reference system
> - distance: computes the distance between two points in meters
> - parseGeoJson: parses [GeoJSON](http://geojson.org/) string
> - parseWkt: parses [WKT](http://en.wikipedia.org/wiki/Well-known_text) string
> - [All geometry relationship functions described here](http://docs.geotools.org/stable/userguide/library/jts/relate.html):
>   - equals
>   - contains
>   - within
>   - overlaps
>   - crosses
>   - touches
>   - intersects
>   - disjoint
>   - relate

Also, because it uses a real geo package, it's probably easily extensible to other spheroids / coordinate systems.

It still uses the original JSON library and has a few other things which could probably stand to be updated, but it seems like a strong starting point.

**Comment by tfmorris** [comment]
I confirmed that the [georefine extension](https://github.com/ryanfb/georefine) still works with the current release. The only minor issue that I noticed is that it breaks the GREL help tab in the expression dialog (#7488), but I've got a prototype of a fix for that.

Given that, I'm not sure why we'd create a competing extension with less functionality.

**Comment by thadguidry** [comment]
@tfmorris because I'm no longer active much with OpenRefine.  It's up to you and other maintainers to make the judgement call.  Yes, I had put in my original comment on this bounty issue to have a look at the georefine extension as you can see.  Likely @SashankBhamidi wanted some Java challenge for themselves and claim the bounty, would be my guess.

Going forward, if you are saying the georefine extension still works, has support for those functions AND MORE, then I'd vote to do whatever makes users comfortable, and at the same time, reduces maintenance burdens for the team.  If that means throwing away @SashankBhamidi efforts then sure, let's do that, no harm, no foul.  The community will only benefit more.

### Questions
1. Are you in communication with @ryanfb who maintains the [georefine extension](https://github.com/ryanfb/georefine)?
2. Who will maintain that [georefine extension](https://github.com/ryanfb/georefine)?
3. Would the community, going forward, be less confused, if our extension list only showed the [georefine extension](https://github.com/ryanfb/georefine) and not @SashankBhamidi geo extension?  (I'm fine with that)

**Comment by SashankBhamidi** [comment]
> Likely @SashankBhamidi wanted some Java challenge for themselves and claim the bounty, would be my guess.

Yes and everything I do involves $WORK

> If that means throwing away @SashankBhamidi efforts then sure, let's do that, no harm, no foul. 
> Would the community, going forward, be less confused, if our extension list only showed the [georefine extension](https://github.com/ryanfb/georefine) and not @SashankBhamidi geo extension? (I'm fine with that)

Let me know what y'all decide, then I'll look into https://github.com/OpenRefine/openrefine-geo-extension/issues/1 accordingly.
